### PR TITLE
Revamp queue system to combine both Chunk Tasks and general Queueing.

### PR DESCRIPTION
**Pull request by dordsor21**
Requires update to WorldEdit (my current open PRs) for the regeneration stuff

### CODE REVIEW

File: Core/src/main/java/com/plotsquared/core/queue/QueueProvider.java [hunk 1]
```diff
@@ -25,24 +25,17 @@
...
-                        e.printStackTrace();
-                        failed = true;
-                    }
-                }
+            @Override public QueueCoordinator getNewQueue(@Nonnull World world) {
                 try {
-                    return (LocalBlockQueue) fallback.getConstructors()[0].newInstance(world);
+                    return (QueueCoordinator) primary.getConstructors()[0].newInstance(world);
                 } catch (Throwable e) {
                     e.printStackTrace();
```
Comment (Sauilitired) [inline Core/src/main/java/com/plotsquared/core/queue/QueueProvider.java:40]: We need better error handling here, and at the very least a more descriptive error (log something)

File: Core/src/main/java/com/plotsquared/core/queue/ChunkQueueCoordinator.java [hunk 2]
```diff
@@ -114,15 +130,15 @@ private void storeCache(final int x, final int y, final int z, final BlockState
...
-    @Override public Location getMax() {
-        return Location.at(getWorld(), top.getX(), top.getY(), top.getZ());
+    @Override @Nonnull public Location getMax() {
+        return Location.at(getWorld().getName(), top.getX(), top.getY(), top.getZ());
     }
 
-    @Override public Location getMin() {
-        return Location.at(getWorld(), bot.getX(), bot.getY(), bot.getZ());
+    @Override @Nonnull public Location getMin() {
+        return Location.at(getWorld().getName(), bot.getX(), bot.getY(), bot.getZ());
```
Comment (Sauilitired) [inline Core/src/main/java/com/plotsquared/core/queue/ChunkQueueCoordinator.java:142]: We should probably add an overloaded location getter for worldedit worlds

File: Core/src/main/java/com/plotsquared/core/queue/ChunkCoordinator.java [hunk 3]
```diff
@@ -25,26 +25,23 @@
...
-
-    public OffsetLocalBlockQueue(LocalBlockQueue parent, int ox, int oy, int oz) {
-        super(parent);
-        this.ox = ox;
-        this.oy = oy;
-        this.oz = oz;
-    }
+import com.plotsquared.core.util.task.PlotSquaredTask;
+
+public abstract class ChunkCoordinator implements PlotSquaredTask {
```
Comment (Sauilitired) [inline Core/src/main/java/com/plotsquared/core/queue/ChunkCoordinator.java:30]: This class needs documentation, as it's the uppermost one. Same with all the methods, as this is part of the API

File: Core/src/main/java/com/plotsquared/core/components/ComponentPresetManager.java [hunk 4]
```diff
@@ -191,11 +188,13 @@
 
                 BackupManager.backup(player, plot, () -> {
                     plot.addRunning();
+                    QueueCoordinator queue = plot.getArea().getQueue();
```
Comment (Sauilitired) [inline Core/src/main/java/com/plotsquared/core/components/ComponentPresetManager.java:191]: I kinda feel like we should have some utility method to simplify this, as this is also done in the set command, and it'd be nice to have it exposed via API. Dunno how tho

File: Bukkit/src/main/java/com/plotsquared/bukkit/generator/DelegatePlotGenerator.java [hunk 5]
```diff
@@ -64,16 +64,15 @@ public DelegatePlotGenerator(ChunkGenerator chunkGenerator, String world) {
...
         Location min = result.getMin();
         int chunkX = min.getX() >> 4;
         int chunkZ = min.getZ() >> 4;
         Random random = new Random(MathMan.pair((short) chunkX, (short) chunkZ));
         try {
             ChunkGenerator.BiomeGrid grid = new ChunkGenerator.BiomeGrid() {
-                @Override
-                public void setBiome(@Range(from = 0, to = 15) int x,
+                @Override public void setBiome(@Range(from = 0, to = 15) int x,
                     @Range(from = 0, to = 15) int z, @Nonnull Biome biome) {
```
Comment (Sauilitired) [inline Bukkit/src/main/java/com/plotsquared/bukkit/generator/DelegatePlotGenerator.java:76]: Get rid of the jb annotations

File: Core/src/main/java/com/plotsquared/core/components/ComponentPresetManager.java (hunk 4 above)
Comment (dordsor21) [reply Core/src/main/java/com/plotsquared/core/components/ComponentPresetManager.java:191]: Technically QueueProvider#of(com.sk89q.worldedit.world.World) but it's easier to use the PlotArea getQueue than giving every class the WorldUtil and then converting a String to a WorldEdit World

Comment (Sauilitired) [reply Core/src/main/java/com/plotsquared/core/components/ComponentPresetManager.java:191]: I meant the entire component setting, not just the queue getter. Should have been clearer about that 🌵

Comment (dordsor21) [reply Core/src/main/java/com/plotsquared/core/components/ComponentPresetManager.java:191]: Ahh right, it's exposed in Plot#setComponent(String component, Pattern blocks, @Nullable QueueCoordinator queue) I guess? If the queue isn't null it writes the blocks to the queue and doesn't enqueue. If the queue is null it makes its own and then enqueues. I'll document that everywhere...

**Pull request merged**